src/prefect/engine/executors/dask.py

```diff
@@ -200,24 +222,71 @@ def start(self) -> Iterator[None]:
 
         Creates a `dask.distributed.Client` and yields it.
         """
-        # import dask client here to decrease our import times
         from distributed import Client
 
         try:
             if self.address is not None:
                 with Client(self.address, **self.client_kwargs) as client:
                     self.client = client
-                    yield self.client
+                    try:
+                        self._pre_start_yield()
+                        yield
+                    finally:
+                        self._post_start_yield()
             else:
                 with self.cluster_class(**self.cluster_kwargs) as cluster:  # type: ignore
                     if self.adapt_kwargs:
                         cluster.adapt(**self.adapt_kwargs)
                     with Client(cluster, **self.client_kwargs) as client:
                         self.client = client
-                        yield self.client
+                        try:
+                            self._pre_start_yield()
+                            yield
+                        finally:
+                            self._post_start_yield()
         finally:
             self.client = None
 
+    def _pre_start_yield(self) -> None:
+        from distributed import Variable
+
+        is_inproc = self.client.scheduler.address.startswith("inproc")  # type: ignore
+        if self.address is not None or is_inproc:
+            self._futures = weakref.WeakSet()
+            self._should_run_var = Variable(
+                f"prefect-{uuid.uuid4().hex}", client=self.client
+            )
+            self._should_run_var.set(True)
+
+    def _post_start_yield(self) -> None:
+        from distributed import wait
+
+        if self._should_run_var is not None:
+            # Multipart cleanup, ignoring exceptions in each stage
+            # 1.) Stop pending tasks from starting
+            try:
+                self._should_run_var.set(False)
+            except Exception:
+                pass
+            # 2.) Wait for all running tasks to complete
+            try:
+                futures = [f for f in list(self._futures) if not f.done()]  # type: ignore
+                if futures:
+                    self.logger.info(
+                        "Stopping executor, waiting for %d active tasks to complete",
+                        len(futures),
+                    )
+                    wait(futures)
+            except Exception:
+                pass
```

Comment:
any benefit to logging the exceptions at warning level?

Reply:
It indicates that things may be slower to shutdown than the user might want. Happy to change to a different level if you want.

Reply:
Makes sense 👍